feat: implement organization ownership transfer (#69)

manab-pr opened the pull request:
feat: implement organization ownership transfer (#69)

  PR Description

  ## Summary

  Implements organization ownership transfer functionality as requested in #69. Organization owners can now transfer ownership to another user
  via the `transferOrganizationOwnership` GraphQL mutation.

  **Two transfer paths:**
  - **Existing member**: Direct role swap - new owner gets OWNER role, previous owner demoted to ADMIN
  - **Non-member**: Ownership transfer invitation sent - transfer completes when invitation is accepted

  ## Changes

  ### GraphQL API
  - Added `transferOrganizationOwnership` mutation in `orgextended.graphql`
  - Returns `OrganizationTransferOwnershipPayload` with organization and invitation status

  ### Database Schema
  - Extended `Invite` schema with `ownership_transfer` boolean field
  - Tracks invitations intended for ownership transfer vs regular membership

  ### Business Logic
  - **Resolver** (`orgextended.resolvers.go`):
    - Verifies caller is current owner
    - Checks if target user is already a member
    - Direct transfer: Updates both roles atomically
    - Invitation path: Creates invite with `ownership_transfer=true`

  - **Invite Hook** (`hooks/invite.go`):
    - Detects `ownership_transfer` flag on invitation acceptance
    - Automatically demotes all current owners to ADMIN role
    - Ensures only one owner exists after transfer

  ### Bug Fixes
  - Fixed duplicate group membership constraint violation in `managedgroups.go`
  - Added existence check before creating group memberships
  - Removed duplicate OWNER enum value in `orgmembership.go`
  - Added OWNER to valid invite role values in `role.go`

  ## Test Plan

  **Unit Tests** (`orgtransfer_test.go`):
  - ✅ `TestTransferOrganizationOwnership_ExistingMember`: Verifies direct transfer
    - Creates owner and member users
    - Transfers ownership
    - Asserts new owner has OWNER role
    - Asserts old owner demoted to ADMIN role

  - ✅ `TestTransferOwnershipViaInvitation`: Verifies invitation path
    - Creates owner and non-member users
    - Sends ownership transfer invitation
    - Accepts invitation
    - Asserts ownership transferred correctly
    - Asserts old owner demoted to ADMIN

  **Manual Testing**:
  - [ ] Transfer to existing admin - verify role change
  - [ ] Transfer to existing member - verify role change
  - [ ] Transfer to non-member - verify invitation sent
  - [ ] Accept ownership invitation - verify transfer completes
  - [ ] Attempt transfer as non-owner - verify rejection
  - [ ] Verify FGA tuples updated correctly
  - [ ] Verify managed groups updated (Admins/Viewers/All Members)

  ## Security Considerations

  - Authorization check: Only current OWNER can initiate transfer
  - Privacy contexts used appropriately for authorized operations
  - FGA tuples automatically updated via existing hooks
  - No breaking changes to existing invite flow

  ## Files Modified

  - `internal/graphapi/schema/orgextended.graphql` - GraphQL schema
  - `internal/graphapi/orgextended.resolvers.go` - Transfer logic
  - `internal/graphapi/query/organization.graphql` - Test queries
  - `internal/ent/schema/invite.go` - Added ownership_transfer field
  - `internal/ent/hooks/invite.go` - Transfer on acceptance
  - `internal/ent/hooks/managedgroups.go` - Fixed duplicate bug
  - `internal/ent/schema/orgmembership.go` - Removed duplicate enum
  - `pkg/enums/role.go` - Added OWNER to valid values
  - `internal/graphapi/orgtransfer_test.go` - Comprehensive tests

  ## Closes

  Closes #69

golanglemonade commented on pkg/enums/role.go:
```diff
@@ -17,9 +17,9 @@ var (
+// Possible default values are "OWNER", "ADMIN", "MEMBER"
 func (Role) Values() (kinds []string) {
-	for _, s := range []Role{RoleAdmin, RoleMember} {
+	for _, s := range []Role{RoleOwner, RoleAdmin, RoleMember} {
```
Ah, I think I see you did the change for invites, however, this shouldn't be changed here because it affects groups as well. Instead you should add the same line that was deleted in orgmembership to invites to add the `Owner` role there.

manab-pr commented:
@golanglemonade  All tests are now passing. Ready for re-review, Please review and let me know if everything looks correct!

golanglemonade commented on internal/graphapi/orgextended.resolvers.go:
```diff
@@ -56,6 +59,124 @@ func (r *mutationResolver) CreateOrganizationWithMembers(ctx context.Context, or
+			orgmembership.OrganizationID(organizationID),
+			orgmembership.UserID(newOwnerID),
+		).
+		First(ctx)
```
I think `Only` is safer, there shouldn't ever be duplicates if DeletedAtIsNil is added to the `Where`
```suggestion
		Only(ctx)
```

golanglemonade commented on internal/graphapi/orgextended.resolvers.go:
```diff
@@ -56,6 +59,124 @@ func (r *mutationResolver) CreateOrganizationWithMembers(ctx context.Context, or
+	if err != nil {
+		log.Error().Err(err).Str("new_owner_id", newOwnerID).Msg("new owner user not found")
+		return nil, newNotFoundError("user")
+	}
```
This should be removed, if the new owner isn't an existing user we should still send them the invitation.

golanglemonade commented on internal/graphapi/orgtransfer_test.go:
```diff
@@ -0,0 +1,154 @@
+)
+
+// TestTransferOrganizationOwnership_ExistingMember tests transferring ownership to an existing org member
+func TestTransferOrganizationOwnership_ExistingMember(t *testing.T) {
```
This doesn't really test the resolvers, instead its just writing the same queries that are in there - which becomes an issue if that code is changed and the paths aren't checked but it looks like you have test coverage.

You should be able to follow a similar test pattern to `TestMutationCreateInvite` to test this new resolver instead.

manab-pr commented:
Thanks for the thorough review! I've addressed all the feedback:

  **Schema Changes:**
  -  Removed `organizationID` parameter - now automatically derived from the authenticated user's JWT context
  -  Changed `newOwnerID: ID!` to `newOwnerEmail: String!` to support inviting users who aren't in the system yet

  **Resolver Updates:**
  -  Now using `GetAuthenticatedUserFromContext()` instead of `GetSubjectIDFromContext()` to get both user ID and organization ID from the
  context
  - Added validation to ensure `organizationID` is present
  -  Removed the user existence check (lines 100-106) - now allows sending invitations regardless of whether the user exists
  -  Changed `First()` to `Only()` for the membership query since there shouldn't be duplicates

  **Test Improvements:**
  -  Completely refactored the test to follow the `TestMutationCreateInvite` pattern
  -  Now uses table-driven tests that actually call the resolver through the GraphQL client
  -  Includes test cases for existing members, non-members, new users, permission denied, and invalid input

@golanglemonade  Please review and let me know if everything looks correct!

golanglemonade commented:
@manab-pr looks like some test failures but the code changes look good:

```

❌  internal/graphapi (3m39.789s) (coverage: 36.8% of statements)
--
  |  
  | DONE 1895 tests, 3 failures in 219.789s
  |  
  | ❌  internal/graphapi (9.894s) (coverage: 1.5% of statements)
  |  
  | DONE 2 runs, 1902 tests, 6 failures in 241.606s
  |  
  | ❌  internal/graphapi (10.309s) (coverage: 1.5% of statements)
  |  
  | DONE 3 runs, 1909 tests, 9 failures in 263.101s
  |  
  | ❌  internal/graphapi (9.86s) (coverage: 1.5% of statements)
  |  
  | === Failed
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership/Transfer_not_owner,_permission_denied (0.02s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership/Transfer_different_org_owner,_no_access (0.07s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership (1.98s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership/Transfer_not_owner,_permission_denied (re-run 1) (0.02s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership/Transfer_different_org_owner,_no_access (re-run 1) (0.06s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership (re-run 1) (1.25s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership/Transfer_not_owner,_permission_denied (re-run 2) (0.02s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership/Transfer_different_org_owner,_no_access (re-run 2) (0.06s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership (re-run 2) (1.35s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership/Transfer_not_owner,_permission_denied (re-run 3) (0.02s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership/Transfer_different_org_owner,_no_access (re-run 3) (0.06s)
  | === FAIL: internal/graphapi TestMutationTransferOrganizationOwnership (re-run 3) (1.20s)


```

golanglemonade commented:
nice, thanks @manab-pr!
